### PR TITLE
Core infrastructure for timer manager debug

### PR DESCRIPTION
**Pull request by muxi**
Add debug infrastructure for timer manager. Part of the effort to investigate the crash in `gpr_cv_wait`.

Note: CV's are struct rather than pointer in iOS, so we are not logging them here.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                            FILE SIZE
 ++++++++++++++ GROWING                              ++++++++++++++
  +0.1%    +544 [None]                                  +712  +0.0%
      +0.1%    +512 [Unmapped]                              +712  +0.0%
      [NEW]      +8 g_fork_count                               0  [ = ]
      [NEW]      +8 g_grpc_debug_timer_manager_stats           0  [ = ]
      [NEW]      +8 g_timer_manager_init_count                 0  [ = ]
      [NEW]      +8 g_timer_manager_shutdown_count             0  [ = ]
      [NEW]      +8 g_timer_wait_cv                            0  [ = ]
      [NEW]      +8 g_timer_wait_err                           0  [ = ]
  +6.7%     +80 src/core/lib/gpr/sync_posix.cc           +80  +6.7%
       +30%     +73 gpr_cv_wait                              +73   +30%
      +6.7%      +7 [Unmapped]                                +7  +6.7%
  +0.6%     +16 src/core/lib/iomgr/timer_manager.cc      +16  +0.6%
      +9.9%     +11 grpc_timer_manager_init                  +11  +9.9%
       +22%     +11 grpc_timer_manager_shutdown              +11   +22%
       +38%      +8 grpc_timer_manager_set_threading          +8   +38%

  +0.0%    +640 TOTAL                                   +808  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,752      Total (>)      1,964,580

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,602      Total (>)     10,886,949

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                            FILE SIZE
 ++++++++++++++ GROWING                              ++++++++++++++
  +0.1%    +592 [None]                                  +856  +0.0%
      +0.1%    +544 [Unmapped]                              +856  +0.0%
      [NEW]      +8 g_fork_count                               0  [ = ]
      [NEW]      +8 g_grpc_debug_timer_manager_stats           0  [ = ]
      [NEW]      +8 g_timer_manager_init_count                 0  [ = ]
      [NEW]      +8 g_timer_manager_shutdown_count             0  [ = ]
      [NEW]      +8 g_timer_wait_cv                            0  [ = ]
      [NEW]      +8 g_timer_wait_err                           0  [ = ]
  +6.7%     +80 src/core/lib/gpr/sync_posix.cc           +80  +6.7%
       +30%     +73 gpr_cv_wait                              +73   +30%
      +6.7%      +7 [Unmapped]                                +7  +6.7%
  +0.6%     +16 src/core/lib/iomgr/timer_manager.cc      +16  +0.6%
      +9.9%     +11 grpc_timer_manager_init                  +11  +9.9%
       +22%     +11 grpc_timer_manager_shutdown              +11   +22%
       +38%      +8 grpc_timer_manager_set_threading          +8   +38%

  +0.0%    +688 TOTAL                                   +952  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,752      Total (>)      1,964,600

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,592      Total (>)     10,886,950

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,469      Total (<)     10,887,471

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,467      Total (<)     10,887,471

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,463      Total (<)     10,887,468

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by AspirinSJL** [comment]
Is this ready for review? Asking because of DO NOT MERGE label.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by muxi** [comment]
@AspirinSJL - removed. It is ready for review

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,470      Total (>)     10,887,469

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/lib/gpr/sync_posix.cc [hunk 1]
```diff
@@ -87,8 +105,34 @@ int gpr_cv_wait(gpr_cv* cv, gpr_mu* mu, gpr_timespec abs_deadline) {
...
+    if (g_grpc_debug_timer_manager_stats) {
+      g_timer_wait_err = err;
+      g_timer_cv_value = (int64_t)cv;
+      g_timer_mu_value = (int64_t)mu;
+      g_grpc_debug_timer_manager_stats(
+          g_timer_manager_init_count, g_timer_manager_shutdown_count,
+          g_fork_count, g_timer_wait_err, g_timer_cv_value, g_timer_mu_value,
+          g_abstime_sec_value, g_abstime_nsec_value);
+    }
+    GPR_ASSERT(err == 0 || err == ETIMEDOUT || err == EAGAIN);
```
Comment (vjpai) [inline src/core/lib/gpr/sync_posix.cc]: Why is this assert inside the `if` ? Instead, you could just do the above work in the `if` and then keep the pre-existing assert outside the `ifdef`, right?

Comment (muxi) [reply src/core/lib/gpr/sync_posix.cc]: Absolutely correct. I just thought it would be the same.

**Comment by muxi** [comment]
It would be a build option turned on by client. I have been checking with the other team to find out what they use for activating this feature (not done yet)

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,466      Total (<)     10,887,469

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +40  +0.0%

  [ = ]       0 TOTAL      +40  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +40  +0.0%

  [ = ]       0 TOTAL      +40  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,463      Total (<)     10,887,470

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,467      Total (<)     10,887,469

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,802      Total (=)      1,964,802

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,468      Total (<)     10,887,471

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +32  +0.0%

  [ = ]       0 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,964,892      Total (=)      1,964,892

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,887,719      Total (>)     10,887,716

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**